[fake_tensor] Move unrecognized_type NotImplemented before ConstProp

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135033

We should not try to do ConstProp on the unrecognized types (e.g. Subclasses).
In case of those types throwing NotImplemented will jump to the next torch_dispatch. 


Test:
```
 python test/functorch/test_aotdispatch.py -k test_aot_test_subclasses_with_tensor_factories
```

IvanKobzarev committed:
[fake_tensor] Move unrecognized_type NotImplemented before ConstProp

[ghstack-poisoned]

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1715,11 +1715,25 @@ def _dispatch_impl(
+        # tensor, it might be related to this line.  Though I'm not sure
+        # how you'll know to read this comment, as this line won't show up
+        # in the stack trace.
+        has_unrecognized_types = _check_for_subclass(flat_args)
+        if has_unrecognized_types:
```
Add a comment that this should be the first thing to happen in FakeTensorMode's torch_dispatch so that people don't end up adding logic above it?

IvanKobzarev committed:
Update on "[fake_tensor] Move unrecognized_type NotImplemented before ConstProp"


We should not try to do ConstProp on the unrecognized types (e.g. Subclasses).
In case of those types throwing NotImplemented will jump to the next torch_dispatch. 


Test:
```
 python test/functorch/test_aotdispatch.py -k test_aot_test_subclasses_with_tensor_factories
```

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[fake_tensor] Move unrecognized_type NotImplemented before ConstProp"


We should not try to do ConstProp on the unrecognized types (e.g. Subclasses).
In case of those types throwing NotImplemented will jump to the next torch_dispatch. 


Test:
```
 python test/functorch/test_aotdispatch.py -k test_aot_test_subclasses_with_tensor_factories
```

[ghstack-poisoned]
